### PR TITLE
[SYCL][ESIMD] Remove wrapping buffer objects into images.

### PR DESCRIPTION
**Pull request by kbobrovs**
This PR removes buffer object wrapping into images, which was used as a SIMD BE workaround and caused a number of problems (e.g. scatter/gather of 1- and 2- byte values worked incorrectly in some cases). It also paves a way for ESIMD/SYCL kernel co-existence in the same source. There are 3 main parts

**Runtime**
- Remove the workaround wrapping.
- Gets rid of majority of __SYCL_EXPLICIT_SIMD__ macro checks in the API code
- Use the new integration header feature - KernelInfo<>::isESIMD() to properly set accessor arguments
in ESIMD kernels.
- Add __init_esimd accessor methods added for FE to pick them up instead of __init for ESIMD kernels, as init sequence is different.

**FE**
- Add 'esimd_acc_ptr' parameter attribute to FE. This attribute is to be used by the FE to mark ESIMD kernel arguments
originating from buffer accessors, which is then translated to BE-specific metadata needed for correct code generation.
- Use 'esimd_acc_ptr' attribute to mark ESIMD kernel pointer arguments originating from accessors
- Use '__init_esimd' initializer in FE for ESIMD kernel accessors.
- Add 'isESIMD()' function to the integration header for the RT to be able to distinguish ESIMD kernels from normal ones for the
  purpose of proper accessor arg setting

**LowerESIMD**
- Add 'buffer_t' metadata for accessor kernel arguments and mark it as surface for proper code gen by ESIMD BE.

### CODE REVIEW

File: sycl/include/CL/sycl/handler.hpp [hunk 1]
```diff
@@ -490,6 +490,8 @@ class __SYCL_EXPORT handler {
     // Empty name indicates that the compilation happens without integration
     // header, so don't perform things that require it.
     if (KI::getName() != nullptr && KI::getName()[0] != '\0') {
+      // TODO support ESIMD in no-integration-header case too.
+      MIsESIMD = KI::isESIMD();
```
Comment (dm-vodopyanov) [inline sycl/include/CL/sycl/handler.hpp:494]: Looks like implementation of isESIMD() is missing. Should it be here or in another PR?

**Review by dm-vodopyanov — COMMENTED** [review]
ABI is broken. Not sure if we can merge any patches with ABI changes now based on https://github.com/intel/llvm/blob/sycl/sycl/doc/ABIPolicyGuide.md#changing-abi. 
@kbobrovs, can this patch be re-written to without changes in ABI?

File: sycl/include/CL/sycl/handler.hpp (hunk 1 above)
Comment (DenisBakhvalov) [reply sycl/include/CL/sycl/handler.hpp:494]: Yes, it's in another PR: https://github.com/intel/llvm/pull/2747/commits/d4f5813b255823f54b6428514b81a1e5be448105#diff-a07ededf6f50832d6f6e956321f97097ff85197e06a3edc9063af88f02d467e2R3836

File: sycl/include/CL/sycl/accessor.hpp [hunk 2]
```diff
@@ -858,27 +858,10 @@ class accessor :
...
-  };
+  ConcreteASPtrType MData;
 
-#ifdef __SYCL_EXPLICIT_SIMD__
-  // TODO In ESIMD accessors usage is limited for now - access range, mem
-  // range and offset are not supported. The cl_mem object allocated for
-  // a global accessor is always wrapped into a 1d image buffer to enable
-  // surface index-based addressing.
-  void __init(OCLImage1dBufferTy ImgBuf) { ImageBuffer = ImgBuf; }
+  const ConcreteASPtrType getNativeImageObj() const { return MData; }
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/include/CL/sycl/accessor.hpp:863]: That is used in ESIMD APIs. Let's rename/refactor as a separate PR

**Comment by kbobrovs** [comment]
@dm-vodopyanov, I removed the ABI breaking change - please take a look.
I had to rebase atop other patches to have a chance to pass testing. Please review
[SYCL][ESIMD] RT: remove wrapping buffer objects into images. … eca13ef
and later

File: sycl/include/CL/sycl/accessor.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/include/CL/sycl/accessor.hpp:863]: I added a TODO

**Comment by kbobrovs** [comment]
Collapsed all commits into the 3 - FE, RT and the LowerESIMD. That's how they should be committed (w/o further squashing)
I think I addressed all review comments.
@Fznamznon - please check/approve the FE (6fb2ce1, 06dd08c)
@DenisBakhvalov - please check/approve the LowerESIMD (873d4a6)
@dm-vodopyanov - please check/approve the RT (055f859)

**Review by dm-vodopyanov — APPROVED** [review]

File: clang/include/clang/Basic/AttrDocs.td [hunk 3]
```diff
@@ -380,6 +380,18 @@ def SYCLRegisterNumDocs : Documentation {
...
 
+def SYCLSimdAccessorPtrDocs : Documentation {
+  let Category = DocCatVariable;
+  let Content = [{
+    The ``__attribute__((esimd_acc_ptr))`` attribute must be used to mark ESIMD
+    kernel pointer parameters which correspond to the original
+    lambda's captured accessors. FE turns the attribute to some metadata
+    required by the ESIMD Back-End.
+    Not supposed to be used directly in the source - SYCL device compiler FE
+    automatically adds it for ESIMD kernels.
```
Comment (Fznamznon) [inline clang/include/clang/Basic/AttrDocs.td:391]: nit: since this attribute is used only internally by fe itself, I would avoid phrase "must be used"
```suggestion
    The ``__attribute__((esimd_acc_ptr))`` attribute is used by FE to mark ESIMD
    kernel pointer parameters which correspond to the original
    lambda's captured accessors. FE turns the attribute to some metadata
    required by the ESIMD Back-End.
    Not supposed to be used directly in the source - SYCL device compiler FE
    automatically adds it for ESIMD kernels.
```

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+// 2) Pointers originating from accessors are marked with 'buffer_t' and proper
+//    argument kind.
+
+#include "Inputs/sycl.hpp"
+
+void test(int val) {
+  cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write> accessorA;
+  cl::sycl::accessor<int, 1, cl::sycl::access::mode::read> accessorB;
+
+  cl::sycl::kernel_single_task<class esimd_kernel>(
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:16]: Could you please use queue::submit and handler::single_task from the mock header instead? And include the mock header as system header using `-internal-isystem` . Like here https://github.com/intel/llvm/blob/sycl/clang/test/CodeGenSYCL/stall_enable.cpp

We need to follow single style in FE test and this style must be as close as possible to the way how we use real sycl headers.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      [=]() __attribute__((sycl_explicit_simd)) {
+        accessorA.use(val);
+        accessorB.use();
+      });
+  // --- Name
+  // CHECK: define spir_kernel void @_ZTSZ4testiE12esimd_kernel(
+  // --- Signature
+  // CHECK:   i32 addrspace(1)* nocapture "VCArgumentDesc"="buffer_t" "VCArgumentIOKind"="0" "VCArgumentKind"="2" %_arg_,
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:24]: Could you please elaborate what is `VCArgumentDesc`, `VCArgumentIOKind` and `VCArgumentKind` and how these things got here?

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+#include "Inputs/sycl.hpp"
+
+void test(int val) {
+  cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write> accessorA;
+  cl::sycl::accessor<int, 1, cl::sycl::access::mode::read> accessorB;
+
+  cl::sycl::kernel_single_task<class esimd_kernel>(
+      [=]() __attribute__((sycl_explicit_simd)) {
+        accessorA.use(val);
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:18]: Could you please also try other kinds of arguments (like POD for example) and check that all metadata is correct.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        accessorB.use();
+      });
+  // --- Name
+  // CHECK: define spir_kernel void @_ZTSZ4testiE12esimd_kernel(
+  // --- Signature
+  // CHECK:   i32 addrspace(1)* nocapture "VCArgumentDesc"="buffer_t" "VCArgumentIOKind"="0" "VCArgumentKind"="2" %_arg_,
+  // CHECK:   i32 "VCArgumentDesc" "VCArgumentIOKind"="0" "VCArgumentKind"="0" %_arg_1,
+  // CHECK:   i32 addrspace(1)* nocapture "VCArgumentDesc"="buffer_t" "VCArgumentIOKind"="0" "VCArgumentKind"="2" %_arg_3)
+  // --- Attributes
+  // CHECK: {{.*}} !kernel_arg_accessor_ptr ![[ACC_PTR_ATTR:[0-9]+]] !sycl_explicit_simd !{{[0-9]+}} {{.*}}{
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:28]: Could you please check that `__init_esimd` is called instead of regular `__init` in case of ESIMD kernel?

File: clang/test/CodeGenSYCL/int_header_esimd.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+// This test checks that
+// 1) New isESIMD() member is generated into the integration header
+// 2) It returns 1 for ESIMD kernels and 0 - for non-ESIMD.
+
+// -- ESIMD Lambda kernel.
+template <typename ID, typename F>
+void kernelA(F f) __attribute__((sycl_kernel)) {
+  f();
+}
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/int_header_esimd.cpp:12]: Same here, please use queue::submit and single_task from the mock header included like system one.

**Review by Fznamznon — COMMENTED** [review]
Is there any chance to merge patches for different components separately? It would simplify "git blaming" in the future.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp (hunk 4 above)
Comment (kbobrovs) [reply clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:16]: I don't see single style in FE tests, so I picked on. But OK, I will try to rework as you suggest.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp (hunk 5 above)
Comment (kbobrovs) [reply clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:24]: This is metadata added by specific Genx pass in the Backendutil.cpp.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp (hunk 4 above)
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:16]: > I don't see single style in FE tests, so I picked on. But OK, I will try to rework as you suggest.

Yes, we don't have single style yet. This problem is already captured, and I remember that someone was working on it, but it is not finished.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp (hunk 5 above)
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:24]: Ah, ok. Thanks.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp (hunk 6 above)
Comment (kbobrovs) [reply clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:18]: Is this a nit? The main point of the test is checking accessors argument translation and MD generation. PODs are no different from other types in this regard - and `int val` should be enough.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp (hunk 7 above)
Comment (kbobrovs) [reply clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:28]: This is basically proven by the fact that an accessor is translated into a single pointer arg (as in __init_esimd) rather than 4 (as in __init). I'll see if I can make direct check. Do you have suggestions, BTW?

File: clang/test/CodeGenSYCL/int_header_esimd.cpp (hunk 8 above)
Comment (kbobrovs) [reply clang/test/CodeGenSYCL/int_header_esimd.cpp:12]: ok

**Comment by kbobrovs** [comment]
> Is there any chance to merge patches for different components separately? It would simplify "git blaming" in the future.

that's the intent. I would like this PR to be committed w/o squashing - as 4 patches.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp (hunk 6 above)
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:18]: My point was to check that we don't generate something unnecessary for non-accessor types in case if it can confuse underlying backend. And I probably missed that you already have captured int, so yes, this comment is more a nit.

File: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp (hunk 7 above)
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:28]: I don't remember exactly, but I was under impression that we don't enable early optimizations for ESIMD mode, am I right?
In case if IR is not optimized, I would just check that there are **two** calls to __init_esimd functions, i.e. (the following LLVM IR may contain errors):
```
CHECK: call spir_func void @{{.*}}__init_esimd{{.*}}(%"class{{.*}}cl::sycl::accessor" addrspace(4)* {{.*}}, i32 addrspace(1)* {{.*}})
```
Or you can even check which argument of the kernel is passed like here **https://github.com/intel/llvm/blob/3b8af2229eeb45f26b610d5e33a007fc6652532e/clang/test/CodeGenSYCL/accessor_inheritance.cpp#L84** .

In case if IR is optimized for ESIMD mode, we can check presence of `__init_esimd` in non-optimized IR.

Comment (kbobrovs) [reply clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp:28]: ok, thanks. Will try

**Comment by kbobrovs** [comment]
@Fznamznon - comments are addressed, please take a look. Had to force-push to maintain the proper patch structure, hope this is OK.
@dm-vodopyanov, please re-approve the RT part.

**Review by Fznamznon — APPROVED** [review]

**Comment by kbobrovs** [comment]
@DenisBakhvalov, please review/approve the LowerESIMD part

**Review by DenisBakhvalov — DISMISSED** [review]

**Comment by kbobrovs** [comment]
@dm-vodopyanov - could you please restore your approval?
@bader, can you please take a look and merge if no objections?

**Review by dm-vodopyanov — APPROVED** [review]

**Comment by bader** [comment]
> @dm-vodopyanov - could you please restore your approval?
> @bader, can you please take a look and merge if no objections?

@kbobrovs, I see a few other PR with similar changes. Could you resolve duplication by closing PRs, which are not intended for merging, please? Make sure that all comments are addressed.
In particular: #2699 and #2748.

I have a couple of comments WRT LLVM pass changes (mentioned in #2748):
1. There is no regression test for LLVM pass
2. I don't quite get the protocol between FE and LLVM pass and I hoped that we can simplify the logic in the pass. Could you answer [this question](https://github.com/intel/llvm/pull/2748#discussion_r521589218), please?

**Comment by kbobrovs** [comment]
@bader - I closed #2747 and #2748. #2699 is not part of this one and still needs to be worked on/merged separately.

> There is no regression test for LLVM pass

There actually is, but it is part of SYCL CodeGen: clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp. 
Will add one more to llvm/test/SYCLLowerIR

> I don't quite get the protocol between FE and LLVM pass and I hoped that we can simplify the logic in the pass. Could you answer this question, please?

sorry, missed that one. The question is:
> Is it possible to have a kernel_arg_accessor_ptr metadata with 0 value?

yes. Only pointer arguments originating from accessors have 1 in the corresponding place in kernel_arg_accessor_ptr metadata

> Make sure that all comments are addressed.

I think all are addressed - reviewers all approved. Do you have some specific  comment in mind (except the above one)?

**Comment by kbobrovs** [comment]
@bader - test is added, please take a look

File: llvm/test/SYCLLowerIR/esimd_acc_ptr.ll [hunk 9]
```diff
@@ -0,0 +1,37 @@
+; RUN: opt -LowerESIMD -S < %s | FileCheck %s
+
+; This test checks that LowerESIMD pass correctly interpretes the
+; 'kernel_arg_accessor_ptr' metadata. Particularly, that it generates additional
+; vector of per-argument metadata (accessible from "genx.kernels" top-level
+; metadata node):
+; - for those arguments having '1' in the corresponding
```
Comment (bader) [inline llvm/test/SYCLLowerIR/esimd_acc_ptr.ll:7]: The code seems to allow any non-zero value. Please, either update the comment or add an assert (or some other check) that FE emitted expected value.

**Review by bader — COMMENTED** [review]
LGTM, except one nit.

File: llvm/test/SYCLLowerIR/esimd_acc_ptr.ll (hunk 9 above)
Comment (kbobrovs) [reply llvm/test/SYCLLowerIR/esimd_acc_ptr.ll]: implemented the nit

**Comment by kbobrovs** [comment]
@bader, please do not squash when merging

**Review by bader — APPROVED** [review]

**Comment by bader** [comment]
> @bader, please do not squash when merging

All other merge options are disabled for this project.
The only way to avoid squash is to create a separate PR for each commit.

**Comment by kbobrovs** [comment]
> All other merge options are disabled for this project.
> The only way to avoid squash is to create a separate PR for each commit.

That's unfortunate. Those PRs won't pass testing. Can we change this setting BTW?
Otherwise, let's squash then, as we don't seem to have a choice. The description can be used as the commit message.
